lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,138 @@
+//===- CPSATSchedulers.cpp - Schedulers using external CPSAT solvers
+//------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of cp-sat programming-based schedulers using external solvers
+// via OR-Tools.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Scheduling/Algorithms.h"
+
+#include "mlir/IR/Operation.h"
+
+#include "ortools/sat/cp_model.h"
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Format.h"
+
+#define DEBUG_TYPE "cpsat-schedulers"
+
+using namespace circt;
+using namespace circt::scheduling;
+using namespace operations_research;
+using namespace operations_research::sat;
+
+using llvm::dbgs;
+using llvm::format;
+
+LogicalResult scheduling::scheduleCPSAT(SharedOperatorsProblem &prob,
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
```

Comment:
(For the other schedulers, the rationale to get `lastOp` as an argument is that the client is expected to know which operation's start time it wants to minimize. In your case, this would save the constraints from all ops to the `makeSpan` variable. I'm fine with both approaches, just don't take an arg that you then ignore.)

Reply:
Right - your take on it is 100% accurate - prior to this iteration, I was buttressing `makeSpan` by `lastOp` but that wasn't working (for whatever reason) so I changed to having `makeSpan` cover all ops. Will drop this.